File: framework/languages/java/e2e/SpotService/Client/src/main/java/systems/zlink/e2e/spotservice/client/Scenarios/SmD8Scenario.java
package systems.zlink.e2e.spotservice.client.Scenarios;

import java.time.Duration;
import java.util.List;
import java.util.UUID;
import systems.zlink.e2e.spotservice.shared.Contracts;
import systems.zlink.stream.connector.ZLinkStreamConnector;

public final class SmD8Scenario extends SpotServiceScenarioContext {
    private SmD8Scenario(SpotServiceScenarioContext context) {
        super(context);
    }

    public static void run(SpotServiceScenarioContext context) {
        new SmD8Scenario(context).execute();
    }

    private void execute() {
        String actorId = "actor-sm-d8-" + UUID.randomUUID().toString().replace("-", "");
        Contracts.ActorProfile profile = new Contracts.ActorProfile("Reconnect", 8, List.of("reconnect"));
        ZLinkStreamConnector first = createStreamConnector(options().streamAEndpoint());
        try {
            first.connect().submit().toCompletableFuture().join();
            Contracts.ActorAuthRes auth = first
                .request(new Contracts.ActorAuthReq(actorId, profile))
                .submit(Contracts.ActorAuthRes.class).toCompletableFuture().join();
            ensure(actorId.equals(auth.actorId()), "SM-D8 initial auth actor mismatch");

            var pending = first
                .request(new Contracts.SlowSessionReq("before-disconnect", 1_000))
                .timeout(Duration.ofSeconds(10))
                .submit(Contracts.SlowSessionRes.class);
            Thread.sleep(100);
            first.close().submit().toCompletableFuture().join();

            boolean pendingFailed = false;
            try {
                pending.toCompletableFuture().join();
            } catch (Exception ignored) {
                pendingFailed = true;
            }
            ensure(pendingFailed, "SM-D8 expected pending request to fail after stream disconnect");
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("stream reconnect scenario interrupted", error);
        } catch (Exception error) {
            throw new IllegalStateException("stream reconnect first phase failed", error);
        } finally {
            closeQuietly(first);
        }

        waitForPlayAEvidence(List.of("StreamDisconnected|play-a|session"));

        ZLinkStreamConnector second = createStreamConnector(options().streamAEndpoint());
        try {
            second.connect().submit().toCompletableFuture().join();
            Contracts.ActorAuthRes auth = second
                .request(new Contracts.ActorAuthReq(actorId, profile))
                .submit(Contracts.ActorAuthRes.class).toCompletableFuture().join();
            ensure(actorId.equals(auth.actorId()), "SM-D8 reauth actor mismatch");

            Contracts.ActorEchoRes reply = second
                .request(new Contracts.ActorEchoReq("after-reconnect", 8, profile))
                .submit(Contracts.ActorEchoRes.class).toCompletableFuture().join();
            ensure(actorId.equals(reply.actorId()), "SM-D8 reconnected actor mismatch");
            ensure("play-a".equals(reply.nodeRid()), "SM-D8 reconnected node mismatch");
            ensure("entry:after-reconnect".equals(reply.value()), "SM-D8 reconnected value mismatch");

            System.out.println("scenario SM-D8 passed");
        } catch (Exception error) {
            throw new IllegalStateException("stream reconnect second phase failed", error);
        } finally {
            closeQuietly(second);
        }

    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Client/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/client/scenarios/SmD8Scenario.kt
package systems.zlink.e2e.kotlin.spotservice.client.scenarios

import systems.zlink.framework.kotlin.*

import java.time.Duration
import kotlinx.coroutines.delay
import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.Env
import systems.zlink.e2e.kotlin.spotservice.client.support.createStreamConnector
import systems.zlink.e2e.kotlin.spotservice.client.support.ensure
import systems.zlink.e2e.kotlin.spotservice.client.support.postJson

internal object SmD8Scenario {
    suspend fun run() {
        val actorId = "actor-sm-d8-reconnect"
        val profile = Contracts.ActorProfile("Reconnect", 8, listOf("reconnect"))
        val first = createStreamConnector(Env.get("e2e.stream.a.endpoint"))
        try {
            first.connect().await()
            val auth = first
                .request(Contracts.ActorAuthReq(actorId, profile))
                .awaitReply<Contracts.ActorAuthRes>()
            ensure(auth.actorId == actorId, "SM-D8 initial auth actor mismatch")

            val pending = first
                .request(Contracts.SlowSessionReq("before-disconnect", 1000))
                .timeout(Duration.ofSeconds(10))
                .submit(Contracts.SlowSessionRes::class.java)
            delay(100)
            first.close().await()

            var pendingFailed = false
            try {
                pending.await()
            } catch (_: Exception) {
                pendingFailed = true
            }
            ensure(pendingFailed, "SM-D8 expected pending request to fail after stream disconnect")
        } finally {
            try {
                first.close().await()
            } catch (_: Exception) {
            }
        }

        postJson(
            Env.get("e2e.http.session.endpoint"),
            "/evidence/wait",
            Contracts.EvidenceWaitReq(
                listOf("ActorEntryDisconnected|session-a|entry|$actorId"),
                10_000,
            ),
            Contracts.EvidenceSnapshot::class.java,
        )

        val second = createStreamConnector(Env.get("e2e.stream.a.endpoint"))
        try {
            second.connect().await()
            val auth = second
                .request(Contracts.ActorAuthReq(actorId, profile))
                .awaitReply<Contracts.ActorAuthRes>()
            ensure(auth.actorId == actorId, "SM-D8 reauth actor mismatch")

            val reply = second
                .request(Contracts.ActorEchoReq("after-reconnect", 8, profile))
                .awaitReply<Contracts.ActorEchoRes>()
            ensure(reply.actorId == actorId, "SM-D8 reconnected actor mismatch")
            ensure(reply.nodeRid == "session-a", "SM-D8 reconnected node mismatch")
            ensure(reply.value == "entry:after-reconnect", "SM-D8 reconnected value mismatch")

            println("scenario SM-D8 passed")
        } finally {
            try {
                second.close().await()
            } catch (_: Exception) {
            }
        }
    }
}
